packages/protocol/test-sol/common/Proxy.t.sol

```diff
@@ -0,0 +1,204 @@
+pragma solidity ^0.5.13;
+
+import "@celo-contracts/common/Proxy.sol";
+
+import "celo-foundry/Test.sol";
+
+import { Constants } from "@test-sol/constants.sol";
+import { Utils } from "@test-sol/utils.sol";
+
+import "@celo-contracts/common/test/GetSetV0.sol";
+import "@celo-contracts/common/test/GetSetV1.sol";
+import "@celo-contracts/common/test/HasInitializer.sol";
+import "@celo-contracts/common/test/MsgSenderCheck.sol";
+
+contract ProxyTest is Test, Constants, Utils {
+  Proxy proxy;
+  GetSetV0 getSet;
+  GetSetV0 proxiedGetSet;
+
+  event ImplementationSet(address indexed implementation);
+  event OwnerSet(address indexed owner);
+
+  function setUp() public {
+    proxy = new Proxy();
+    getSet = new GetSetV0();
+    proxiedGetSet = new GetSetV0();
+  }
+}
+
+contract ProxyTest_getOwner is ProxyTest {
+  function test_ShouldGetTheAddressOfTheOwner() public {
+    assertEq(proxy._getOwner(), address(this));
+  }
+}
+
+contract ProxyTest_setImplementation is ProxyTest {
+  function test_ShouldAllowTheOwnerToSetAnImplementation() public {
+    proxy._setImplementation(address(getSet));
+    assertEq(proxy._getImplementation(), address(getSet));
+  }
+
+  function test_shouldNotAllowANonOwnerToSetAnImplementation() public {
+    vm.expectRevert("sender was not owner");
+    vm.prank(msg.sender);
+    proxy._setImplementation(address(getSet));
+  }
+
+  function test_shouldAllowTheImplementationToBeUpdated() public {
+    proxy._setImplementation(address(getSet));
+    GetSetV1 getSet1 = new GetSetV1();
+    proxy._setImplementation(address(getSet1));
+    assertEq(proxy._getImplementation(), address(getSet1));
+  }
+
+  function test_ShouldNotAffectLogicRelatedStorage() public {
+    uint256 numberTotest = 42;
+    proxy._setImplementation(address(getSet));
+    proxiedGetSet.set(numberTotest);
+    GetSetV1 getSet1 = new GetSetV1();
+    proxy._setImplementation(address(getSet1));
+
+    assertEq(proxiedGetSet.get(), numberTotest);
+  }
+
+  function test_ShouldEmitImplementationSet() public {
```

Comment:
Ditto

Reply:
what would be you're preference for renaming this function?

Reply:
same format as the reverting test

`test_Emits_EVENTNAME`